Documentation for creating a new cluster on a different AWS account (#728)
kubernetes-sigs/cluster-api-provider-aws: docs/roleassumption.md

# Creating clusters using cross account role assumption using kiam

> This document outlines the list of steps to create the target cluster via cross account role assumption using KIAM. 

Reviewer: suggest linking `KIAM` to https://github.com/uswitch/kiam

KIAM lets the controller pod(s) to assume an AWS role that enables them create AWS resources necessary to create an
operational cluster. This way we wouldn't have to mount any AWS credentials or load environment variables to 
supply AWS credentials to the CAPA controller. This is automatically taken care by the KIAM components.
Note: If you dont want to use KIAM and rather want to mount the credentials as secrets, you may still achieve cross 
account role assumption by using multiple profiles. 

### Glossory

* Target Account - The account where the cluster is created
* Source Account - The AWS account where the CAPA controller runs. 

## Assumptions

> 1. The CAPA controllers are running in 1 AWS account and you want to create the target cluster in another AWS account.

Reviewer: ```suggestion
1. The CAPA controllers are running in an AWS account and you want to create the target cluster in another AWS account.
```

> (We could also use this doc to create a cluster in the same account)

Reviewer 2: While technically true, there's no need for this in a single account, right? I recommend removing this line, to keep it as simple as possible for the reader.

2. This assumes that you start with no existing clusters. 

## High level steps

1. Creating a bootstrap/management cluster in AWS - This can be done by running the phases in clusterctl
    * Uses the existing provider components yaml
2. Setting up cross account roles

> 3. Deploying the Kiam server/agent

Reviewer 2: Above you wrote KIAM, here you wrote Kiam, below you wrote kiam.I don't know which is correct, but please make it consistent.

4. Create the target cluster (through kiam)
    * Uses different provider components with no secrets and annotation to indicate the IAM Role to assume. 

## 1. Creating the bootstrap cluster in AWS
Using clusterctl command we can create a new cluster in AWS which in turn will act as the 
bootstrap cluster to create the target cluster(in a different AWS account. This can be achieved by using the phases in 
clusterctl to perform all the steps except the pivoting. This will provide us with a bare-bones functioning cluster that 
we can use as a bootstrap cluster. 
To begin with follow the steps in this getting started guide
(https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/docs/getting-started.md) to setup the environment

> except for creating the actual cluster. Instead follow the steps below to create the cluster.

Reviewer: suggest listing the sections that need to be followed.

create a new cluster using kind for bootstrapping purpose by running:

> ```$xslt

Reviewer: I don't know what the `$xslt`. Can you please tell me what that is?

Author: It should have been $sh. $xslt was added as the default code for mark up. Its a XML format and used to convert one xml document into another.

> kind create cluster

Reviewer: ```suggestion
kind create cluster --name=<Cluster-Name>
```

```
and get its kube config path by running
```
export KIND_KUBECONFIG=`kind get kubeconfig-path`
```

Use the following commands to create new (bootstrap) cluster in AWS
```
kubectl alpha phases apply-cluster-api-components --provider-components cmd/clusterctl/examples/aws/out/provider-components.yaml \
--kubeconfig $KIND_KUBECONFIG

kubectl alpha phases apply-cluster --cluster cmd/clusterctl/examples/aws/out/cluster.yaml --kubeconfig $KIND_KUBECONFIG

kubectl alpha phases apply-machines --machines cmd/clusterctl/examples/aws/out/machines.yaml 

> --kubeconfig $KIND_KUBECONFIG

Author: good points!!
I can say-  remove the definition for the node in the machines.yaml so that we we only create the control plane ? 
Sure I will move the create machines after apply addons.

kubectl alpha phases get-kubeconfig --provider aws --cluster-name <CLUSTER_NAME> --kubeconfig $KIND_KUBECONFIG

> export AWS_KUBECONFIG=`pwd`/kubeconfig

Reviewer: ```suggestion
export AWS_KUBECONFIG=$PWD/kubeconfig
```

> kubectl alpha phases apply-addons -a cmd/clusterctl/examples/aws/out/addons.yaml 

Reviewer: ```suggestion
kubectl alpha phases apply-addons -a cmd/clusterctl/examples/aws/out/addons.yaml --kubeconfig $AWS_KUBECONFIG
```

Reviewer: and remove line 61

--kubeconfig $AWS_KUBECONFIG

```

Verify that all the pods in the kube-system namespace are running smoothly. Also you may remove the additional node in 
the machines example yaml since we are only interested in running the controllers that runs in control plane node
(although its not required to make any changes there). You can destroy your local kind cluster by running 
```$xslt

> make kind-reset

Reviewer 3: this will delete the kind cluster named `clusterapi` but the above instructions create a kind cluster with the default name, `kind`

```

> ## 2. Setting up cross account roles

Reviewer: ```suggestion
## 2. Setting up cross account IAM roles
```

> In this step we will new roles/policy in total across 2 different AWS accounts.

Reviewer 3: In this step we will `<accidentally a word>` new roles/policy

> First lets start by creating the roles in the account where the AWS controller runs. Following the directions 

Reviewer 3: ```suggestion
First let's start by creating the roles in the account where the AWS controller runs. Following the directions 
```

> posted here:https://github.com/uswitch/kiam/blob/master/docs/IAM.md create a "kiam_server" role

Reviewer 3: ```suggestion
[posted here](https://github.com/uswitch/kiam/blob/master/docs/IAM.md) create a "kiam_server" role
```

in AWS that only has a single managed policy with a single permission "sts:AssumeRole". Also add a trust policy on the 
 "kiam_server" role to include the role attached to the Control plane instance as a trusted entity. This looks something
 like this:

>  ```$xslt

Reviewer 3: ```suggestion

 ```json
```

{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::<AWS_ACCOUNT_NUMBER>:role/control-plane.cluster-api-provider-aws.sigs.k8s.io"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

Next we must establish a link between this "kiam_server" role on source AWS account and the role on target AWS account that has the permissions to create new cluster.
Begin by running the clusterawsadm cli to create a new stack on the target account where the target cluster is created. Make sure you use the credentials for target AWS account for running this step.

> ```clusterawsadm alpha bootstrap create-stack```

Reviewer 3: ```suggestion

```clusterawsadm alpha bootstrap create-stack```

```

> Then sign-in to the target AWS account to establish the link as mentioned above. Create a new Role with the permission policy set to "controllers.cluster-api-provider-aws.sigs.k8s.io". Lets name this role "cluster-api" for future reference. Add a new trust relationship to include the "kiam_server" role from the source account as trusted entity. This is shown below:

Reviewer 3: ```suggestion
Then sign-in to the target AWS account to establish the link as mentioned above. Create a new Role with the permission policy set to "controllers.cluster-api-provider-aws.sigs.k8s.io". Lets name this role "cluster-api" for future reference. Add a new trust relationship to include the "kiam_server" role from the source account as trusted entity. This is shown below:

```

> ```

Reviewer 3: ```suggestion
```json
```

{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::<SOURCE_AWS_ACCOUNT_NUMBER>:role/kserver"
      },
      "Action": "sts:AssumeRole"
    }
  ]

> }

Reviewer 3: ```suggestion
}

```

```
## 3. Deploying the Kiam server & agent

> By Now, your target cluster must be up & running. Make sure your KUBECONFIG pointing to the cluster in the target account. 

Reviewer 2: nit: Now -> now
pointing -> is pointing (or points, your choice)

> Create new secrets using the steps outlined here: https://github.com/uswitch/kiam/blob/master/docs/TLS.md 

Reviewer: ```suggestion
Create new secrets using the steps outlined [here] (https://github.com/uswitch/kiam/blob/master/docs/TLS.md) 
```

Apply the manifest shown below: 
Make sure you update the argument to include your source AWS account "--assume-role-arn=arn:aws:iam::<SOURCE_AWS_ACCOUNT>:role/kiam_server"

> server.yaml

Reviewer 3: ```suggestion
server.yaml

```

          livenessProbe:
            exec:
              command:
              - /kiam
              - health
              - --cert=/etc/kiam/tls/server.pem
              - --key=/etc/kiam/tls/server-key.pem
              - --ca=/etc/kiam/tls/ca.pem
              - --server-address=127.0.0.1:443
              - --gateway-timeout-creation=1s
              - --timeout=5s
            initialDelaySeconds: 10
            periodSeconds: 10
            timeoutSeconds: 10
          readinessProbe:
            exec:
              command:
              - /kiam
              - health
              - --cert=/etc/kiam/tls/server.pem
              - --key=/etc/kiam/tls/server-key.pem
              - --ca=/etc/kiam/tls/ca.pem
              - --server-address=127.0.0.1:443
              - --gateway-timeout-creation=1s
              - --timeout=5s
            initialDelaySeconds: 3
            periodSeconds: 10
            timeoutSeconds: 10
---
apiVersion: v1
kind: Service
metadata:
  name: kiam-server
  namespace: kube-system
spec:
  clusterIP: None
  selector:
    app: kiam
    role: server
  ports:
  - name: grpclb
    port: 443
    targetPort: 443
    protocol: TCP
```

> agent.yaml 

Reviewer 3: ```suggestion

agent.yaml 

```

              add: ["NET_ADMIN"]
          image: quay.io/uswitch/kiam:master # USE A TAGGED RELEASE IN PRODUCTION
          imagePullPolicy: Always
          command:
            - /kiam
          args:
            - agent
            - --iptables
            - --host-interface=cali+
            - --json-log
            - --port=8181
            - --cert=/etc/kiam/tls/agent.pem
            - --key=/etc/kiam/tls/agent-key.pem
            - --ca=/etc/kiam/tls/ca.pem
            - --server-address=kiam-server:443
            - --prometheus-listen-addr=0.0.0.0:9620
            - --prometheus-sync-interval=5s
            - --gateway-timeout-creation=1s
          env:
            - name: HOST_IP
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
          volumeMounts:
            - mountPath: /etc/ssl/certs
              name: ssl-certs
            - mountPath: /etc/kiam/tls
              name: tls
            - mountPath: /var/run/xtables.lock
              name: xtables
          livenessProbe:
            httpGet:
              path: /ping
              port: 8181
            initialDelaySeconds: 3
            periodSeconds: 3
```

> server-rbac.yaml

Reviewer 3: ```suggestion

server-rbac.yaml

```

```
---
kind: ServiceAccount
apiVersion: v1
metadata:
  name: kiam-server
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: kiam-read
rules:
- apiGroups:
  - ""
  resources:
  - namespaces
  - pods
  verbs:
  - watch
  - get
  - list
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: kiam-read
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: kiam-read
subjects:
- kind: ServiceAccount
  name: kiam-server
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: kiam-write
rules:
- apiGroups:
  - ""
  resources:
  - events
  verbs:
  - create
  - patch
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: kiam-write
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: kiam-write
subjects:
- kind: ServiceAccount
  name: kiam-server
  namespace: kube-system
```

> After Deploying the above components make sure that the kiam_server & kiam_agent pods are up & running. 

Reviewer 2: nit: Deploying -> deploying

## 4. Create the target cluster
Make sure you create copy of the "aws/out" directory called "out2". To create the target cluster we must update the provider_components.yaml generated in the out2 directory. 
1. Remove the credentials secret added at the bottom of the provider_components.yaml and do not mount the secret 
2. Add the following annotation to the template of aws-provider-controller-manager stateful set to specify the new role that was created in target account.
```
      annotations:
        iam.amazonaws.com/role: arn:aws:iam::<TARGET_AWS_ACCOUNT>:role/cluster-api
```
3. Also add this below annotation to the "aws-provider-system" namespace 
```
  annotations:
    iam.amazonaws.com/permitted: ".*"
```

Create a new cluster using the steps similar to the one used to create the source cluster. They are as follows:
export SOURCE_KUBECONFIG=<PATH_TO_SOURCE_CLUSTER_KUBECONFIG>
```

> kubectl alpha phases apply-cluster-api-components --provider-components cmd/clusterctl/examples/aws/out2/provider-components.yaml \

Reviewer 3: these bash commands are a little inconsistent, sometimes there is a `\` linebreak and sometimes there is not. Might be easier to read / copy paste  with consistent formatting.

Author: Added the missing line breaks.

# Creating clusters using cross account role assumption using KIAM

This document outlines the list of steps to create the target cluster via cross account role assumption using [KIAM]https://github.com/uswitch/kiam. 
KIAM lets the controller pod(s) to assume an AWS role that enables them create AWS resources necessary to create an
operational cluster. This way we wouldn't have to mount any AWS credentials or load environment variables to 
supply AWS credentials to the CAPA controller. This is automatically taken care by the KIAM components.
Note: If you dont want to use KIAM and rather want to mount the credentials as secrets, you may still achieve cross 
account role assumption by using multiple profiles. 

### Glossory

* Target Account - The account where the cluster is created

> * Source Account - The AWS account where the CAPA controller runs. 

Reviewer 3: would it be useful to call this out as a management cluster?

## Assumptions

> 1. The CAPA controllers are running in an AWS account and you want to create the target cluster in another AWS account.

Reviewer 3: This is more of a result of the steps instead of a prior assumption, maybe this should be under a #goals section?

2. This assumes that you start with no existing clusters. 

## High level steps

1. Creating a bootstrap/management cluster in AWS - This can be done by running the phases in clusterctl
    * Uses the existing provider components yaml
2. Setting up cross account IAM roles
3. Deploying the KIAM server/agent
4. Create the target cluster (through KIAM)
    * Uses different provider components with no secrets and annotation to indicate the IAM Role to assume. 

> ## 1. Creating the bootstrap cluster in AWS

Reviewer 3: in this section 'bootstrap cluster' should be referenced as a 'management cluster' since it will be the owner of the children clusters. The term bootstrap cluster is generally reserved for the ephemeral cluster that exists before a management cluster.

Begin by running the clusterawsadm cli to create a new stack on the target account where the target cluster is created. Make sure you use the credentials for target AWS account before creating the stack.

```clusterawsadm alpha bootstrap create-stack```

Then sign-in to the target AWS account to establish the link as mentioned above. Create a new Role with the permission policy set to "controllers.cluster-api-provider-aws.sigs.k8s.io". Lets name this role "cluster-api" for future reference. Add a new trust relationship to include the "kiam_server" role from the source account as trusted entity. This is shown below:

In "controllers.cluster-api-provider-aws.sigs.k8s.io" role(target AWS account)

```$json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::<SOURCE_AWS_ACCOUNT_NUMBER>:role/kserver"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

## 3. Deploying the KIAM server & agent
By now, your target cluster must be up & running. Make sure your KUBECONFIG pointing to the cluster in the target account. 

> Create new secrets using the steps outlined [here]https://github.com/uswitch/kiam/blob/master/docs/TLS.md 

Reviewer 3: The markdown syntax for links is `[text](url)`

          livenessProbe:
            httpGet:
              path: /ping
              port: 8181
            initialDelaySeconds: 3
            periodSeconds: 3
```

server-rbac.yaml

```
---
kind: ServiceAccount
apiVersion: v1
metadata:
  name: kiam-server
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: kiam-read
rules:
- apiGroups:
  - ""
  resources:
  - namespaces
  - pods
  verbs:
  - watch
  - get
  - list
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: kiam-read
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: kiam-read
subjects:
- kind: ServiceAccount
  name: kiam-server
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: kiam-write
rules:
- apiGroups:
  - ""
  resources:
  - events
  verbs:
  - create
  - patch
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: kiam-write
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: kiam-write
subjects:
- kind: ServiceAccount
  name: kiam-server
  namespace: kube-system
```

> After deploying the above components make sure that the kiam_server & kiam_agent pods are up & running. 

Reviewer 3: ```suggestion
After deploying the above components make sure that the kiam_server and the kiam_agent pods are up and healthy.
```

## 4. Create the target cluster

> Make sure you create copy of the "aws/out" directory called "out2". To create the target cluster we must update the provider_components.yaml generated in the out2 directory. 

Reviewer 3: think it's worth adding a command to help with this instruction? this feels slightly complex, but I think it's fine for a first draft

3. Deploying the KIAM server/agent
4. Create the target cluster (through KIAM)
    * Uses different provider components with no secrets and annotation to indicate the IAM Role to assume. 

## 1. Creating the management cluster in AWS
Using clusterctl command we can create a new cluster in AWS which in turn will act as the 
management cluster to create the target cluster(in a different AWS account. This can be achieved by using the phases in 
clusterctl to perform all the steps except the pivoting. This will provide us with a bare-bones functioning cluster that 
we can use as a management cluster. 
To begin with follow the steps in this [getting started guide](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/docs/getting-started.md) to setup the environment
except for creating the actual cluster. Instead follow the steps below to create the cluster.

create a new cluster using kind for bootstrapping purpose by running:

```$sh
kind create cluster --name <CLUSTER_NAME>
```

and get its kube config path by running

```$sh
export KIND_KUBECONFIG=`kind get kubeconfig-path`
```

Use the following commands to create new management cluster in AWS. 
```$sh

> kubectl alpha phases apply-cluster-api-components --provider-components cmd/clusterctl/examples/aws/out/provider-components.yaml \

Reviewer 4: ```suggestion
clusterctl alpha phases apply-cluster-api-components --provider-components cmd/clusterctl/examples/aws/out/provider-components.yaml \
```

--kubeconfig $KIND_KUBECONFIG

> kubectl alpha phases apply-cluster --cluster cmd/clusterctl/examples/aws/out/cluster.yaml --kubeconfig $KIND_KUBECONFIG

Reviewer 4: ```suggestion
clusterctl alpha phases apply-cluster --cluster cmd/clusterctl/examples/aws/out/cluster.yaml --kubeconfig $KIND_KUBECONFIG
```

```

We only need to create the control plane on the cluster running in AWS source account. Since the example includes definition for a worker node, you may delete it. 

```$sh

> kubectl alpha phases apply-machines --machines cmd/clusterctl/examples/aws/out/machines.yaml --kubeconfig $KIND_KUBECONFIG

Reviewer 4: ```suggestion
clusterctl alpha phases apply-machines --machines cmd/clusterctl/examples/aws/out/machines.yaml --kubeconfig $KIND_KUBECONFIG
```

> kubectl alpha phases get-kubeconfig --provider aws --cluster-name <CLUSTER_NAME> --kubeconfig $KIND_KUBECONFIG

Reviewer 4: ```suggestion
clusterctl alpha phases get-kubeconfig --provider aws --cluster-name <CLUSTER_NAME> --kubeconfig $KIND_KUBECONFIG
```

export AWS_KUBECONFIG=$PWD/kubeconfig

> kubectl alpha phases apply-addons -a cmd/clusterctl/examples/aws/out/addons.yaml --kubeconfig $AWS_KUBECONFIG

Reviewer 4: ```suggestion
clusterctl alpha phases apply-addons -a cmd/clusterctl/examples/aws/out/addons.yaml --kubeconfig $AWS_KUBECONFIG
```

          livenessProbe:
            exec:
              command:
              - /kiam
              - health
              - --cert=/etc/kiam/tls/server.pem
              - --key=/etc/kiam/tls/server-key.pem
              - --ca=/etc/kiam/tls/ca.pem
              - --server-address=127.0.0.1:443
              - --gateway-timeout-creation=1s
              - --timeout=5s
            initialDelaySeconds: 10
            periodSeconds: 10
            timeoutSeconds: 10
          readinessProbe:
            exec:
              command:
              - /kiam
              - health
              - --cert=/etc/kiam/tls/server.pem
              - --key=/etc/kiam/tls/server-key.pem
              - --ca=/etc/kiam/tls/ca.pem
              - --server-address=127.0.0.1:443
              - --gateway-timeout-creation=1s
              - --timeout=5s
            initialDelaySeconds: 3
            periodSeconds: 10
            timeoutSeconds: 10
---
apiVersion: v1
kind: Service
metadata:
  name: kiam-server
  namespace: kube-system
spec:
  clusterIP: None
  selector:
    app: kiam
    role: server
  ports:
  - name: grpclb
    port: 443
    targetPort: 443
    protocol: TCP
```

agent.yaml 

```

> apiVersion: extensions/v1beta1

Reviewer 4: Same as above regarding the api version referenced

## 4. Create the target cluster
Make sure you create copy of the "aws/out" directory called "out2". To create the target cluster we must update the provider_components.yaml generated in the out2 directory as shown below (to be run from the repository root directory):

```
cp cmd/clusterctl/examples/aws/out cmd/clusterctl/examples/aws/out2
vi cmd/clusterctl/examples/aws/out2/provider-components.yaml
```

1. Remove the credentials secret added at the bottom of the provider_components.yaml and do not mount the secret 
2. Add the following annotation to the template of aws-provider-controller-manager stateful set to specify the new role that was created in target account.
```
      annotations:
        iam.amazonaws.com/role: arn:aws:iam::<TARGET_AWS_ACCOUNT>:role/cluster-api
```
3. Also add this below annotation to the "aws-provider-system" namespace 
```
  annotations:
    iam.amazonaws.com/permitted: ".*"
```

Create a new cluster using the steps similar to the one used to create the source cluster. They are as follows:
```
export SOURCE_KUBECONFIG=<PATH_TO_SOURCE_CLUSTER_KUBECONFIG>

> kubectl alpha phases apply-cluster-api-components --provider-components cmd/clusterctl/examples/aws/out2/provider-components.yaml \

Reviewer 4: Instead of deploying the provider components a second time, it might be better to just edit the deployed provider components:

```sh
kubectl -n aws-provider-system edit statefulset aws-provider-controller-manager
# manually remove the referenced secret and save
kubectl -n aws-provider-system delete secret <secret name>
```

This will make it more clear the changes being made to the provider components, and it will also avoid leaving the bootstrapping secret in the deployed cluster, since updating the provider components will not delete the secret.

Author: I am not sure if I follow you correctly. The provider components is deployed for the second time on a different cluster running in AWS. The first time its deployed on the local cluster created by kind.

Reviewer 4: doh, that is correct. My apologies.

--kubeconfig $SOURCE_KUBECONFIG

> kubectl alpha phases apply-cluster --cluster cmd/clusterctl/examples/aws/out2/cluster.yaml --kubeconfig $SOURCE_KUBECONFIG

Reviewer 4: This should be `clusterctl` instead of `kubectl`. That said, I don't think we should necessarily use `clusterctl` here.

The correct `kubectl` based command is:
```sh
kubectl --kubeconfig $SOURCE_KUBECONFIG apply -f cmd/clusterctl/examples/aws/out2/cluster.yaml
```

> kubectl alpha phases apply-machines --machines cmd/clusterctl/examples/aws/out2/machines.yaml \

Reviewer 4: `clusterctl` instead of `kubectl`. That said, I don't think we should necessarily use `clusterctl` here.

The correct `kubectl` based command is:
```sh
kubectl --kubeconfig $SOURCE_KUBECONFIG apply -f cmd/clusterctl/examples/aws/out2/machines.yaml
```

--kubeconfig $SOURCE_KUBECONFIG

> kubectl alpha phases get-kubeconfig --provider aws --cluster-name <TARGET_CLUSTER_NAME> --kubeconfig $SOURCE_KUBECONFIG

Reviewer 4: ```suggestion
clusterctl alpha phases get-kubeconfig --provider aws --cluster-name <TARGET_CLUSTER_NAME> --kubeconfig $SOURCE_KUBECONFIG
```

export TARGET_KUBECONFIG=`pwd`/kubeconfig

> kubectl alpha phases apply-addons -a cmd/clusterctl/examples/aws/out2/addons.yaml \

Reviewer 4: This should be `clusterctl` instead of `kubectl`. That said, I don't think we should necessarily use `clusterctl` here.

The correct `kubectl` based command is:
```sh
kubectl --kubeconfig $SOURCE_KUBECONFIG apply -f cmd/clusterctl/examples/aws/out2/addons.yaml
```

Author: good catch!
